pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -0,0 +1,219 @@
+		if timeout.Before(timenow) {
+			return ErrTimedout
+		}
+	}
+
+	// After a positive timeout check we need to update the timeout and
+	// schedule an update so that we can either set or update the Timeout
+	a.tokenChannel <- tokenData{name, token.ClientName, timenow, timeout}
+	return nil
+}
+
+// updates known clients timeouts, removing clients if they go away
+// and calculate the flush period based on actual timeouts, with a
+// 10 minutes default
+func (a *TokenTimeoutUpdater) updateClientTimeouts() time.Duration {
+	// Default to 10 minutes
+	var flush int32 = 600
+
+	// Update known ones only
+	if len(a.clientTimeouts) > 0 {
+		for name := range a.clientTimeouts {
+			c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					delete(a.clientTimeouts, name)
+				}
+			}
+			if c.AccessTokenTimeoutSeconds == nil {
+				delete(a.clientTimeouts, name)
+			} else {
+				a.clientTimeouts[name] = *c.AccessTokenTimeoutSeconds
+				// Use half period if it is smaller
+				half := int32(math.Ceil(float64(a.clientTimeouts[name]) / 2))
+				if flush > half {
+					flush = half
+				}
+			}
+		}
+	}
+	return time.Duration(flush) * time.Second
+}
+
+func (a *TokenTimeoutUpdater) clientTimeout(name string) int32 {
+	t, ok := a.clientTimeouts[name]
+	if !ok {
+		c, err := a.oauthClient.Get(name, metav1.GetOptions{})
+		if err != nil {
+			if kerrors.IsNotFound(err) {
+				glog.Info("Token timeout set but OAuthClient %s not found!", name)
+			} else {
+				glog.Error("Failed to fetch OAuthClient for timeout value")
+			}
+			return 0
+		}
+		if c.AccessTokenTimeoutSeconds == nil {
+			glog.Info("Token timeout set but OAuthClient %s has no timeout!", name)
+			return 0
+		}
+		t = *c.AccessTokenTimeoutSeconds
+		a.clientTimeouts[name] = t
+	}
+	return t
+}
+
+func (a *TokenTimeoutUpdater) insert(td tokenData) {
+	a.data[td.name] = td
+	a.tree.ReplaceOrInsert(&tokenDataRef{td.name, td.timeout})
+}
+
+func (a *TokenTimeoutUpdater) remove(td tokenData) {
+	a.tree.Delete(&tokenDataRef{td.name, td.timeout})
+	delete(a.data, td.name)
+}
+
+func (a *TokenTimeoutUpdater) flush(flushHorizon time.Time) {
+	flushedTokens := 0
+	totalTokens := len(a.data)
+
+	glog.Infof("Flushing tokens timing out before %v", flushHorizon)
+
+	for {
+		item := a.tree.Min()
+		if item == nil {
+			// out of items
+			break
+		}
+		td := a.data[item.(*tokenDataRef).name]
+		if td.timeout.Before(flushHorizon) {
+			delta := a.clientTimeout(td.client)
+			// ignore tokens for which we can't find a delta
+			if delta > 0 {
+				token, err := a.tokens.Get(td.name, metav1.GetOptions{})
+				if err == nil { // just ignore if the token is gone
+					newtimeout := int32(td.seen.Sub(token.CreationTimestamp.Time)/time.Second) + delta
+					if newtimeout > token.TimeoutsIn {
+						token.TimeoutsIn = newtimeout
+						// ignore errors, if it fails to update there isn't much we can do
+						_, _ = a.tokens.Update(token)
```

Comment:
Can we issue a patch and avoid the `Get` in this logic?

Log errors.

Reply:
Maybe, I'll look into that.

Reply:
Seem Patch using JSON Patch time will work.